MoltenVK/MoltenVK/Utility/MVKEnvironment.h

```diff
@@ -73,6 +73,19 @@
 #	define MVK_SUPPORT_IOSURFACE_BOOL (__TV_OS_VERSION_MIN_REQUIRED >= __TVOS_11_0)
 #endif
 
+/**
+ * Enable use of private Metal APIs.
+ *
+ * Enabling this build setting during a MoltenVK build will allow MoltenVK to
+ * extend its functionality by using certain private Metal API calls, but it 
+ * will also disqualify the app from being distributed via Apple App Stores.
+ *
+ * Disabled by default.
+ */
+#ifndef MVK_USE_METAL_PRIVATE_API
+#	define MVK_USE_METAL_PRIVATE_API		0
+#endif
+
```

Comment:
I had actually envisioned a second setting that can be set at run time and controls whether Metal SPIs would be used at run time.

Reply:
Interesting. What's the use case for that? And how would it work (given that the `setLineWidth:` must be completely stripped from any binaries hitting the App Store)?

Reply:
(Huh, I can't see @spnda's comment...)

IIRC, many Objective-C symbols aren't imported the way regular symbols are. Class descriptors and non-fragile i-var descriptors are, but protocol descriptors, property descriptors, and method selectors aren't. Ultimately, due to the dynamic nature of Objective-C, every Objective-C method and property access resolves to a call to `objc_msgSend()`. It would thus not be possible to ascertain which Objective-C interfaces we are using by inspecting the `dyld` info. That can only be determined at run time, when the methods and properties are actually used.

The other reason I wanted a separate switch for run time is that I was afraid that using Metal SPIs might cause other problems, and having that switch would make it easier to test that without having to recompile MoltenVK.

Reply:
> (Huh, I can't see @spnda's comment...)

Neither can I. I assume they must have deleted it.

> It would thus not be possible to ascertain which Objective-C interfaces we are using by inspecting the `dyld` info. That can only be determined at run time, when the methods and properties are actually used.

Are you referring to the App Store figuring out what calls we're making?

If so, we ran into a private-API issue early on with `IOSurface` on iOS. Are you thinking the `IOSurface` class was undefined, and the App Store could tell we were using it, but the `setLineWidth:` method would not be traceable? 

I can't say yes or no to that until we try, but even so, I'm wondering if it might be traceable if it's a unique symbol.

Regardless, I definitely want to avoid a cat-and-mouse game with Apple about what they can or cannot detect in their App Store filtering. I'd rather be clean about it, and ensure we only send legal builds to the App Store.

How about we get this build-time level of functionality in, and then explore and test options for adding runtime sneakiness in afterwards?

Reply:
> Are you referring to the App Store figuring out what calls we're making?

Yes.

> If so, we ran into a private-API issue early on with `IOSurface` on iOS. Are you thinking the `IOSurface` class was undefined, and the App Store could tell we were using it, but the `setLineWidth:` method would not be traceable?

Yes.



> Regardless, I definitely want to avoid a cat-and-mouse game with Apple about what they can or cannot detect in their App Store filtering. I'd rather be clean about it, and ensure we only send legal builds to the App Store.
> 
> How about we get this build-time level of functionality in, and then explore and test options for adding runtime sneakiness in afterwards?

Very well. I actually wanted both methods anyway. I'm sorry I wasn't clear about that earlier.